whelktool/scripts/cleanups/2022/06/lxl-3880-shuffle-work-titles.groovy

```diff
@@ -0,0 +1,400 @@
+import whelk.filter.LanguageLinker
+import whelk.util.Statistics
+import whelk.util.Unicode
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+moved = getReportWriter('moved.tsv')
+blankLangRemoved = getReportWriter('blank-lang-removed.tsv')
+extraLang = getReportWriter('lang-added.tsv')
+propertyAlreadyExists = getReportWriter('property-already-exists.tsv')
+brokenLinks = getReportWriter('broken-links.tsv')
+multiple = getReportWriter('multiple.txt')
+
+languageLinker = buildLanguageMap()
+Map a = (Map<String, List>) languageLinker.ambiguousIdentifiers
+ambiguousLangNames = a.keySet()
+ambiguousLangIds = a.values().flatten()
+
+languageNames = languageLinker.map.keySet() + languageLinker.substitutions.keySet() + languageLinker.ambiguousIdentifiers.keySet()
+
+INSTANCE_OF = "instanceOf"
+HAS_PART = "hasPart"
+HAS_TITLE = "hasTitle"
+EXPRESSION_OF = "expressionOf"
+TRANSLATION_OF = "translationOf"
+LANGUAGE = "language"
+ID = '@id'
+
+selectByCollection('bib') { bib ->
+    Map instance = bib.graph[1]
+    def instanceOf = instance.find { it.key == INSTANCE_OF }
+    def shortId = bib.doc.shortId
+
+    if (!instanceOf)
+        return
+
+    if (multipleValues(instanceOf.value)) {
+        multiple.println("$shortId\t$INSTANCE_OF")
+        return
+    }
+
+    Map work = asList(instanceOf.value)[0]
+
+    // Remove superfluous blank languages
+    def worklang = asList(work[LANGUAGE])
+    def whichOne = worklang.findAll { it[ID] in ambiguousLangIds }
+    def blankLang = worklang.find { !it[ID] }
+    if (blankLang) {
+        def mapped = mapBlankLanguages([blankLang], whichOne)
+        if ((worklang - blankLang).toSet() == mapped.toSet()) {
+            worklang.remove(blankLang)
+            blankLangRemoved.println([shortId, blankLang.label, worklang].join('\t'))
+        }
+    }
+
+    def expressionOf = work.find { it.key == EXPRESSION_OF }
+
+    if (expressionOf) {
+        if (multipleValues(expressionOf.value)) {
+            multiple.println("$shortId\t$EXPRESSION_OF")
+            return
+        }
+
+        // Move expressionOf content primarily to translationOf, otherwise to instanceOf
+        def target = work.find { it.key == TRANSLATION_OF } ?: instanceOf
+
+        Map hub = asList(expressionOf.value)[0]
+        if (hub[ID]) {
+            expressionOf.value = loadThing(hub[ID])
+            if (!expressionOf.value) {
+                brokenLinks.println([shortId, hub[ID]].join('\t'))
```

Comment:
> I think it was a bit hard to follow the script. The giant if-else structure could be improved.
> 
> See comments, below, otherwise ship

Agree that the structure is messy (would be much easier if it wasn't for the detailed reporting). _Tiny_ improvement in [3d2293a](https://github.com/libris/librisxl/pull/1130/commits/3d2293a30d7cbece0b12adbc2bdf9e99c6f7bd7a).